gobblin-service/src/main/java/org/apache/gobblin/service/modules/core/state/GobblinServiceState.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.service.modules.core.state;
+
+
+public abstract class GobblinServiceState {
+  public abstract boolean isNewStateAllowed(GobblinServiceState state);
+  public boolean isTypeOf (Class candidate) {
+    return this.getClass().getName().equals(candidate.getName());
+  }
+
+  public String toString() {
+    return this.getClass().getName();
+  }
+
+  public static class GobblinServiceOffline extends GobblinServiceState {
+    @Override
+    public boolean isNewStateAllowed(GobblinServiceState state) {
+      return state instanceof GobblinServiceOffline
+          || state instanceof GobblinServiceBecomingStandby
+          || state instanceof GobblinServiceBecomingActive;
+    }
+  }
+
+  public static class GobblinServiceActive extends GobblinServiceState {
+    @Override
+    public boolean isNewStateAllowed(GobblinServiceState state) {
+      return state instanceof GobblinServiceBecomingStandby
+          || state instanceof GobblinServiceActive
+          || state instanceof GobblinServiceBecomingOffline;
+    }
+  }
+
+  public static class GobblinServiceStandby extends GobblinServiceState {
+    @Override
+    public boolean isNewStateAllowed(GobblinServiceState state) {
+      return state instanceof GobblinServiceBecomingActive
+          || state instanceof GobblinServiceStandby
+          || state instanceof GobblinServiceBecomingOffline;
+    }
+  }
+
+  public static class GobblinServiceBecomingStandby extends GobblinServiceState {
```

Comment:
Why are these intermediate states relevant in GobblinService? Isn't active/standby/offline sufficient to capture all possible states of a service manager?

Reply:
Also, is the eventual goal that we have a global state checker that looks at the aggregate state of all Gobblin service managers and detect any state violations (e.g. no two nodes can be leaders at the same time) similar to Helix?

Reply:
When serviceManager receives a restli request, we might want different behaviors between transient state and a stable state. For example, if a service is in Standby state, the behavior might be "forwarding the message to active node for processing". However if a service is in BecomingStandby state, the behavior might be "Reject the user request" or "Caching the requests for future replay". Depending on what the service have and how we want to define our system, we might need different states for different things. Having a transient state helps:
1) Differentiate behaviors
2)Justify current system state, reject any illegal state transition.

As for if we need a global state checker, I assume Helix can provide a safety here, which means no two leaders at the same time. However we still provide our own state machine so that even if Helix state machine diverged from us, we can still justify ourselves and provide a source of truth why state transition happened.